lib/opal/cli_runners/firefox.rb

```diff
@@ -0,0 +1,356 @@
+                return path.endsWith('/index.#{ext}') ? {
+                  url: '/index.map', map: #{map.to_json}
+                } : null;
+              }
+            });
+            </script>
+          </head><body>
+            <script src='/index.#{ext}'#{module_type}></script>
+          </body></html>
+        HTML
+      end
+
+      def firefox_host
+        ENV['CHROME_HOST'] || DEFAULT_CHROME_HOST
+      end
+
+      def firefox_port
+        ENV['CHROME_PORT'] || DEFAULT_CHROME_PORT
+      end
+
+      def with_firefox_server
+        if firefox_server_running?
+          yield
+        else
+          run_firefox_server { yield }
+        end
+      end
+
+      def run_firefox_server
+        raise 'Firefox server can be started only on localhost' if firefox_host != DEFAULT_CHROME_HOST
+
+        profile = mktmpprofile
+        # For options see https://github.com/puppeteer/puppeteer/blob/main/packages/puppeteer-core/src/node/FirefoxLauncher.ts
+        firefox_server_cmd = %{#{OS.shellescape(firefox_executable)} \
+          --no-remote \
+          --profile #{profile} \
+          --headless \
+          --remote-debugging-port #{firefox_port} \
+          #{ENV['FIREFOX_OPTS']}}
+
+        firefox_pid = Process.spawn(firefox_server_cmd, in: OS.dev_null, out: OS.dev_null, err: OS.dev_null)
+
+        Timeout.timeout(30) do
+          loop do
+            break if firefox_server_running?
+            sleep 0.5
+          end
+        end
+
+        yield
+      rescue Timeout::Error
+        puts 'Failed to start firefox server'
+        puts 'Make sure that you have it installed and that its version is > 100'
+        exit(1)
+      ensure
+        if OS.windows? && firefox_pid
+          Process.kill('KILL', firefox_pid) unless system("taskkill /f /t /pid #{firefox_pid} >NUL 2>NUL")
+        elsif firefox_pid
+          Process.kill('HUP', firefox_pid)
+        end
+        FileUtils.rm_rf(profile) if profile
+      end
+
+      def firefox_server_running?
+        puts "Connecting to #{firefox_host}:#{firefox_port}..."
+        TCPSocket.new(firefox_host, firefox_port).close
+        # Firefox CDP endpoints are initialized after the CDP port is ready
+        # this causes first requests to fail
+        # wait until the CDP endpoints are ready
+        response = Net::HTTP.get_response('localhost', '/json/list', firefox_port)
+        raise Errno::EADDRNOTAVAIL if response.code != '200'
+        true
+      rescue Errno::ECONNREFUSED, Errno::EADDRNOTAVAIL
+        false
+      end
+
+      def firefox_executable
+        ENV['MOZILLA_FIREFOX_BINARY'] ||
+          if OS.windows?
+            [
+              'C:/Program Files/Mozilla Firefox/firefox.exe'
+            ].each do |path|
+              next unless File.exist? path
+              return path
+            end
+          elsif OS.macos?
+            '/Applications/Firefox.app/Contents/MacOS/Firefox'
+          else
+            %w[
+              firefox
```

Comment:
Debian users have firefox-esr by default, may be a good idea to add it as well

Reply:
done